compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/generators/DelegatedMemberGenerator.kt

```diff
@@ -0,0 +1,191 @@
+ */
+internal class DelegatedMemberGenerator(
+    private val components: Fir2IrComponents
+) : Fir2IrComponents by components {
+
+    // Generate delegated members for [subClass]. The synthetic field [irField] has the super interface type.
+    fun generateDelegatedMembers(irField: IrField, subClass: IrClass) {
+        (irField.type as IrSimpleTypeImpl)?.classOrNull?.owner?.declarations?.filter {
+            it.isDelegatable() && !subClass.declarations.any { decl -> isOverriding(irBuiltIns, decl, it) }
+        }?.map {
+            if (it is IrSimpleFunction) {
+                val function = generateDelegatedFunction(subClass, irField, it)
+                subClass.addMember(function)
+            } else if (it is IrProperty) {
+                generateDelegatedProperty(subClass, irField, it)
+            }
+        }
+    }
+
+    private fun IrDeclaration.isDelegatable(): Boolean {
+        // TODO: replace checks on descriptors
+        val descriptor = this.descriptor
+        return descriptor is CallableMemberDescriptor && descriptor.isOverridable
+                && (!isFakeOverride || !DescriptorUtils.getAllOverriddenDescriptors(descriptor).all { DescriptorUtils.isClass(it.containingDeclaration) })
+                && !(this is IrSimpleFunction && hasDefaultImplementation())
+    }
+
+    private fun IrSimpleFunction.hasDefaultImplementation(): Boolean {
+        var realFunction: IrSimpleFunction? = this
+        while (realFunction != null && realFunction.isFakeOverride) {
+            realFunction = realFunction.overriddenSymbols?.firstOrNull()?.owner
+        }
+        return realFunction != null
+                && (realFunction.modality != Modality.ABSTRACT && realFunction.origin == IrDeclarationOrigin.IR_EXTERNAL_JAVA_DECLARATION_STUB
+                || realFunction.annotations.hasAnnotation(FqName("kotlin.jvm.JvmDefault")))
+    }
+
+    private fun generateDelegatedFunction(subClass: IrClass, irField: IrField, superFunction: IrSimpleFunction): IrSimpleFunction {
+        val startOffset = irField.startOffset
+        val endOffset = irField.endOffset
+        val descriptor = WrappedSimpleFunctionDescriptor()
+        val origin = IrDeclarationOrigin.DELEGATED_MEMBER
+        val delegateFunction = symbolTable.declareSimpleFunction(descriptor) { symbol ->
+            IrFunctionImpl(
+                startOffset,
+                endOffset,
+                origin,
+                symbol,
+                superFunction.name,
+                superFunction.visibility,
+                if (superFunction.modality == Modality.ABSTRACT) Modality.OPEN else superFunction.modality,
+                superFunction.returnType,
+                superFunction.isInline,
+                superFunction.isExternal,
+                superFunction.isTailrec,
+                superFunction.isSuspend,
+                superFunction.isOperator,
+                superFunction.isExpect
+            ).apply {
```

Comment:
Since you're not using the default factory defined in `SymbolTable.declare*`, I'd like to double-check whether we won't need to set `@Metadata`. Please refer to https://github.com/JetBrains/kotlin/pull/3261

Reply:
Not sure about this: I did use SymbolTable.declareSimpleFunction?

Reply:
Yes, you're using `SymbolTable.declareSimpleFunction` with your own `functionFactory`: https://github.com/JetBrains/kotlin/blob/master/compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/SymbolTable.kt#L745

Okay, now there is no more default function factory at all. So you should set metadata info by yourself, like: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrDeclarationStorage.kt#L482

When I mentioned my previous PR, around that time, `SymbolTable.declare*`s have default IR declaration factory where metadata is set. But, in `fir2ir`, we almost always pass our own factory without metadata setting. So, that's why I wondered the use of `declareSimpleFunction` with own function factory still misses metadata setting. Since default function factory is gone, it seems it's up to caller's responsibility to property set metadata source.

Reply:
done.